staticTmpDir: enable forcing tmpDir to a static path

gergelytraveltime opened the pull request:
Added staticTmpDir option to Config.
This will force tmpDir to be created at a static path. Tmp dir will be autodeleted at exit, maybe that
`staticTmpDirLeaveAfterError` was defined, then we'll show a warning that it needs to be manually cleaned up.

This helps with using stryker4s with Bazel, as Bazel uses the path to the workspace (folder) to determine where to store the local cache (hashes the path to create the sandbox's name).
Enabled not autoremoving the folder to be able to debug issues (as Bazel support is early-stage).

hugo-vrijswijk commented:
Thanks for the PR! This would also close #668 I believe. Could you name the option the same as in StrykerJS? So `clean-temp-dir` with default false (except on error is a nice touch 👌)? Then it could also be enabled for the 'normal' temp directory setup.

hugo-vrijswijk commented on core/src/main/scala/stryker4s/run/MutantRunner.scala:
```diff
@@ -69,11 +69,38 @@ class MutantRunner(
+      val staticTmpDir = targetDir / "stryker4s-tmpDir"
+      Resource
+        .eval(Files[IO].createDirectory(staticTmpDir).map(_ => staticTmpDir))
+        .onFinalizeCase(staticTmpDirFinalizeCase(staticTmpDir))
```
I think what you want here is `Resource.makeCase` with the first argument being creating the directory, and the second handling deleting it when the resource is done

gergelytraveltime replied:
Done.

hugo-vrijswijk commented:
> > Thanks for the PR! This would also close #668 I believe. Could you name the option the same as in StrykerJS? So `clean-temp-dir` with default false (except on error is a nice touch ok_hand)? Then it could also be enabled for the 'normal' temp directory setup.
> 
> I guess `clean-temp-dir` with default `true`, right?

Yes! Sorry, I didn't sleep much last night haha

hugo-vrijswijk commented on core/src/main/scala/stryker4s/run/MutantRunner.scala:
```diff
@@ -69,11 +69,38 @@ class MutantRunner(
+      Files[IO].deleteRecursively(staticTmpDir)
+    case _: Resource.ExitCase.Errored =>
+      // Enable the user do some manual actions on the staticTmpDir before she retries.
+      if (config.staticTmpDirLeaveAfterError) {
```
`Files[IO].createTempDirectory` should do the trick

gergelytraveltime replied:
Ah, I found it, `createTempDirectory`

gergelytraveltime commented:
> Added staticTmpDir option to Config. This will force tmpDir to be created at a static path. Tmp dir will be autodeleted at exit, maybe that `staticTmpDirLeaveAfterError` was defined, then we'll show a warning that it needs to be manually cleaned up.
> 
> This helps with using stryker4s with Bazel, as Bazel uses the path to the workspace (folder) to determine where to store the local cache (hashes the path to create the sandbox's name). Enabled not autoremoving the folder to be able to debug issues (as Bazel support is early-stage).

Updated description:

    staticTmpDir: enable forcing tmpDir to a static path
    
    Added staticTmpDir option to Config.
    This will force tmpDir to be created at a static path.
    
    This helps with using stryker4s with Bazel, as Bazel uses the path to the
    workspace (folder) to determine where to store the local cache (hashes
    the path to create the sandbox's name).
    
    -----
    
    Enabled not autoremoving the folder to be able to debug issues:
    
    Tmp dir will be autodeleted at exit, maybe that
    `clearTmpDir` was disabled, then we'll show a warning that
    it needs to be manually cleaned up.
    On error the tmpDir is not deleted (whether or not static).

hugo-vrijswijk commented on core/src/main/scala/stryker4s/run/MutantRunner.scala:
```diff
@@ -69,11 +69,41 @@ class MutantRunner(
+  private def prepareTmpDir(targetDir: Path): Resource[IO, Path] = {
+    val tmpDirCreated = if (config.staticTmpDir) {
+      val staticTmpDir = targetDir / "stryker4s-tmpDir"
+      Files[IO].createDirectory(staticTmpDir).map(_ => staticTmpDir)
```
```suggestion
      Files[IO].createDirectory(staticTmpDir).as(staticTmpDir)
```

hugo-vrijswijk commented on core/src/main/scala/stryker4s/config/Config.scala:
```diff
@@ -21,7 +21,9 @@ final case class Config(
     concurrency: Int = Config.defaultConcurrency,
-    debug: DebugOptions = DebugOptions()
+    debug: DebugOptions = DebugOptions(),
+    staticTmpDir: Boolean = false,
```
Could you also add a test for the new options in `ConfigReaderTest` and add docs for them in `docs/configuration.md`? Then I think we can merge this!

gergelytraveltime replied:
Done.

hugo-vrijswijk reviewed (commented):
Looks good!! One minor style thing, and a question about tests and docs for the new config options, but then we can merge this

hugo-vrijswijk commented on core/src/main/scala/stryker4s/run/MutantRunner.scala:
```diff
@@ -69,11 +69,41 @@ class MutantRunner(
+    Resource.makeCase(tmpDirCreated)(tmpDirFinalizeCase)
+  }
+
+  private val tmpDirFinalizeCase: (Path, Resource.ExitCase) => IO[Unit] = {
```
Is there a specific reason you made this a `val` instead of a `def`? I think it would be slightly cleaner as a def since you wouldn't have to have the `case`

gergelytraveltime replied:
I think I still need to have the case.

hugo-vrijswijk replied:
Yes, but couldn't it be `def tmpDirFinalizeCase(exitCase....): IO[Unit]`?

gergelytraveltime replied:
Switched to a def in any case.

gergelytraveltime replied:
> I think I still need to have the case.

I think what I don't need to have is a `match`, but it's also not necessary when defining as a `val`.

hugo-vrijswijk replied:
Like I said, I didn't sleep much last night, but i see now why the `case` is there

hugo-vrijswijk commented on docs/configuration.md:
```diff
@@ -120,6 +120,28 @@ Set the Scala dialect that should be used for parsing Scala files. The default i
+### `static-tmp-dir` [`boolean`]
+
+**Config file:** `static-tmp-dir: true`
+**Default value:** `false`
```
```suggestion
**Default value:** `false`
**Since:** `v0.15.0`  
```

gergelytraveltime replied:
Ok.

hugo-vrijswijk commented on docs/configuration.md:
```diff
@@ -120,6 +120,28 @@ Set the Scala dialect that should be used for parsing Scala files. The default i
+### `clean-tmp-dir` [`boolean`]
+
+**Config file:** `clean-tmp-dir: false`
+**Default value:** `true`
```
```suggestion
**Default value:** `true`
**Since:** `v0.15.0`  
```

gergelytraveltime replied:
Ok.

gergelytraveltime commented:
Applied requested docs change and added new tests for `MutantRunner`.

gergelytraveltime commented on core/src/test/scala/stryker4s/run/MutantRunnerTest.scala:
```diff
@@ -95,5 +107,104 @@ class MutantRunnerTest extends Stryker4sIOSuite with MockitoIOSuite with LogMatc
+        verify(reporterMock, times(1)).onRunFinished(captor.capture)
+        val runReport = captor.value.report.files.loneElement
+
+        "Setting up mutated environment..." shouldBe loggedAsInfo
```
Is adding such simple assertions here ok? Or test for even less, as mutation are checked in another test case?

hugo-vrijswijk replied:
To keep the test clean, it's best to only test for the changes that this test has compared to the others. So since logging is already covered in other tests you don't need to assert for it again

gergelytraveltime replied:
Removed some assertions. Should I remove even more?

gergelytraveltime commented:
Anything else necessary @hugo-vrijswijk?

hugo-vrijswijk reviewed (approved):
Thanks so much for this PR and the clean solution!

gergelytraveltime commented:
> Thanks so much for this PR and the clean solution! I'll merge it now, and if you really can't wait, you can use a snapshot release soon. Otherwise this will be in the next release (which shouldn't be too far away)

Thanks for all your help and assistance!
